networkx/algorithms/traversal/breadth_first_search.py

```diff
@@ -370,6 +371,53 @@ def bfs_successors(G, source, depth_limit=None, sort_neighbors=None):
     yield (parent, children)
 
 
+def bfs_layers(G, sources):
+    """Returns an iterator of all the layers in breadth-first search traversal.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        A graph over which to find the layers using breadth first search.
+
+    sources : list of nodes in `G`
+        Specify starting nodes for single source or multiple sources breadth-first search
+
+    Yields
+    -------
+    layer: list of nodes
+        Yields list of nodes at the same distance from sources
+
+    Examples
+    --------
+    >>> G = nx.path_graph(5)
+    >>> dict(enumerate(nx.bfs_layers(G, [0, 4])))
+    {0: [0, 4], 1: [1, 3], 2: [2]}
+    >>> H = nx.Graph()
+    >>> H.add_edges_from([(0, 1), (0, 2), (1, 3), (1, 4), (2, 5), (2, 6)])
+    >>> dict(enumerate(nx.bfs_layers(H, [1])))
+    {0: [1], 1: [0, 3, 4], 2: [2], 3: [5, 6]}
+    >>> dict(enumerate(nx.bfs_layers(H, [1, 6])))
+    {0: [1, 6], 1: [0, 3, 4, 2], 2: [5]}
+    """
+    for source in list(sources):
+        if not G.has_node(source):
+            raise nx.NetworkXError(f"The node {source} is not in the graph.")
+
+    current_layer = list(sources)
+    visited = set(sources)
+
+    # this is basically BFS, except that the current layer only stores the nodes at
+    while current_layer:
+        next_layer = list()
+        for node in current_layer:
+            for child in G[node]:
+                if child not in visited:
+                    visited.add(child)
+                    next_layer.append(child)
+        yield current_layer
```

Comment:
We should `yield` before the loop.   ...   Why? because we know the value of `current_layer` so we can yield. And we might not be asked to visit the next layer (the user might stop the loop by not calling "next" or stopping their loop over the layers).  In that case, we will have saved the time spent generating `next_layer`.

So, I would move the "yield" statement to immediately after the `while` statement.